research-software-directory/RSD-as-a-service
# 879 - public profile page

# Public profile page (using opt-in approach)

Closes #879 
Closes #1017 

Changes proposed in this pull request:
* User can link his/her ORCID to RSD account and enable public profile in the user settings
* User can unlink ORCID authentication method (only after public profile page is disabled)
* This PR includes lot of changes that should be tested.       
* **In order to test this feature you need valid ORCID account and ORCID authentication method defined in .env file**.  If you use default RSD development .env file the ORCID login method is already defined and enabled. You can create ORCID account for free.
* Added user settings section to documentation
* Minor unrelated fixes to contributor show all/less button and software/project card layout

**We should also test locally with the real RSD data from the backup in order to identify possible corner cases**.

How to test:
* Add value to .env file in the ORCID authentication section (see .env.example)
```.env
# consumed by: authentication, frontend/utils/loginHelpers
ORCID_REDIRECT_COUPLE=http://www.localhost/auth/couple/orcid
```
* `make start` to build app an create test items
* Test all login methods and confirm that these work. I refactored props in the login methods to exclude unused claims.
* Login as rsd_user using SURFconext as `professor1` for example
* Navigate to user settings. You will see a list of authentication methods and `Link your ORCID account` button (see image 1).
* Link your ORCID account. It should appear in the list of authentication methods. Note! The ORCID is not your real instead it is one provided by ORCID test api. Copy this ORCID as we will use it later.
* Accept all terms and enable public profile (see image 2)
* Create software and add yourself as contributor. Note! You need to change the ORCID because for authentication we use test api of ORCID. When you link ORCID you are linking it with test ORCID. Go back to settings and copy the ORCID from the account and use this ORCID for the contributor.
* View software page. In the list of the contributors you should see link next to contributor name. This link will bring you to Public page of this person. (see image 3)
* Create project and add team member in the same way you dit it with software contributor. The list of team members who enabled public profile should also have a link to profile page.
* Now logout as professor1. Navigate to software and project where you added yourself and confirm that link to public profile page is still present and accessible to all RSD visitors
* Next, try to login with ORCID. This should work. Navigate to settings and disable public profile page. Note that the delete button for ORCID authentication method is enabled. You are able delete ORCID authentication method. Before removing ORCID validate that link to public profile page is removed from your contributor item on software page and team member on the project page.
* Try forcing public profile page by using url `http://localhost/profile/{YOUR ORCID}/software`. This should result in 404.
* Enable public profile page in settings and confirm that you can access the page.
* Repeat these tests with orcid linking as rsd_admin. If you use same .env settings as I use, professor3 is defined as rsd_admin

## User settings - authentication methods
![image](https://github.com/research-software-directory/RSD-as-a-service/assets/9204081/b4f1dcaf-ffdb-4d9f-823b-a4289222b4ea)

## User settings - enable public profile
![image](https://github.com/research-software-directory/RSD-as-a-service/assets/9204081/1b934757-b7bf-404b-abfd-357062e38ab4)

## Link to public profile page (link icon shown next to name)
![image](https://github.com/research-software-directory/RSD-as-a-service/assets/9204081/965f55f8-fcc9-44d3-a37d-5b426df20b56)

## Public profile page
![image](https://github.com/research-software-directory/RSD-as-a-service/assets/9204081/a27e4eea-ac21-464e-bea4-1c4bfc09b81f)

## Empty public profile page
![image](https://github.com/research-software-directory/RSD-as-a-service/assets/9204081/0c3efe02-f5f6-45e3-836d-b80ae25e3aeb)


PR Checklist:
